File: Classes/AtlasLoader.cpp
//
//  AtlasLoader.cpp
//  FlappyBird
//
//  Created by Meone on 16/7/16.
//
//

#include "AtlasLoader.h"

AtlasLoader *AtlasLoader::sharedAtlasLoader = nullptr;

AtlasLoader *AtlasLoader::getInstance()
{
    if (sharedAtlasLoader == NULL) {
        sharedAtlasLoader = new AtlasLoader();
        if (!sharedAtlasLoader->init()) {
            delete sharedAtlasLoader;
            sharedAtlasLoader = NULL;
            CCLOG("ERROR: Could not init sharedAtlasLodaer");
        }
    }
    return sharedAtlasLoader;
}

void AtlasLoader::destroyInstance()
{
    CC_SAFE_DELETE(sharedAtlasLoader);
}

AtlasLoader::AtlasLoader(){}

bool AtlasLoader::init()
{
    return true;
}

void AtlasLoader::loadAtlas(string filename)
{
    auto textureAtlas = Director::getInstance()->getTextureCache()->addImage("atlas.png");
    this->loadAtlas(filename, textureAtlas);
}

void AtlasLoader::loadAtlas(string filename, cocos2d::Texture2D *texture)
{
    string data = FileUtils::getInstance()->getStringFromFile(filename);
    unsigned int pos;
    Atlas atlas;
    pos = data.find_first_of("\n");
    string line = data.substr(0, pos);
    data = data.substr(pos + 1);
    while(line != "") {
        sscanf(line.c_str(), "%s %d %d %f %f %f %f",
               atlas.name, &atlas.width, &atlas.height,
               &atlas.start.x, &atlas.start.y,
               &atlas.end.x, &atlas.end.y);
        atlas.start.x = 1024 * atlas.start.x;
        atlas.start.y = 1024 * atlas.start.y;
        atlas.end.x = 1024 * atlas.end.x;
        atlas.end.y = 1024 * atlas.end.y;
        
        pos = data.find_first_of("\n");
        line = data.substr(0, pos);
        data = data.substr(pos + 1);
        
        if (atlas.name == string("land")) {
            atlas.start.x += 1;
            
        }
        Rect rect = Rect(atlas.start.x, atlas.start.y, atlas.width, atlas.height);
        auto frame = SpriteFrame::createWithTexture(texture, rect);
        this->_spriteFrames.insert(string(atlas.name), frame);
    }
}

SpriteFrame *AtlasLoader::getSpriteFrameByName(string name)
{
    return this->_spriteFrames.at(name);
}

File: src/app/tools/AtlasLoader.lua
--print("AtlasLoader")
local _atlasData = {}

local _atlasLoader = nil
local imageName = "atlas.png"
local fileName = "atlas.txt"

local AtlasLoader = class("AtlasLoader", cc.Object)

--AtlasLoader._index = AtlasLoader

--setmetatable(_atlasLoader, AtlasLoader)

function AtlasLoader:getInstance()
    if _atlasLoader == nil then
        _atlasLoader = AtlasLoader:create()
    end
    return _atlasLoader
end

function AtlasLoader:destroyInstance()
	_atlasLoader = nil
end

--生成sprite lua file
local function generateSpriteFrameCache()
	-- print("AtlasLoader generateSpriteFrameCache")
	local atlasFileData = cc.FileUtils:getInstance():getStringFromFile(fileName)
	local pos = string.find(atlasFileData, "\n")
	local line = string.sub(atlasFileData, 0, pos)
	local atlasFileData = string.sub(atlasFileData, pos+1, string.len(atlasFileData))
	local tmpSprite
	while (line ~= "")
	do
		tmpSprite = {}
		for i in string.gmatch(line, "%S+") do
			tmpSprite[#tmpSprite+1] = i
		end
		local xSprite = math.floor(1024 * (tonumber(tmpSprite[4])) + 0.1)
		local ySprite = math.floor(1024 * (tonumber(tmpSprite[5])) + 0.1)
		if(tmpSprite[1] == "land") then
			xSprite = xSprite + 1
		end
		_atlasData[tmpSprite[1]] = { width = tmpSprite[2], height = tmpSprite[3], x = xSprite, y = ySprite }
		--计算下一行,找不到为nil
		pos = string.find(atlasFileData, "\n")
		if(pos == nil) then
			break
		end
		line = string.sub(atlasFileData, 0, pos)
		atlasFileData = string.sub(atlasFileData, pos+1, string.len(atlasFileData))
	end
end

function AtlasLoader:ctor()
	generateSpriteFrameCache()
end

function AtlasLoader:getSpriteFrameByName(name)
	local textureAtlas = cc.Director:getInstance():getTextureCache():addImage(imageName)
	return cc.SpriteFrame:createWithTexture(textureAtlas, _atlasData[name])
end

return AtlasLoader
